Port analysis/test to SystemScalarConverter

jwnimmer-tri opened the pull request:
FYI The Doxygen at http://drake.mit.edu/doxygen_cxx/group__system__scalar__conversion.html and its additions merged in #6743 (which will reach the website this weekend) helps explain what's going on here.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/7031)
<!-- Reviewable:end -->

jwnimmer-tri commented:
+@edrumwri for feature review, please -- you seem to be the primary contributor to these files, so I want to make sure you understand and agree with the changes.

+@EricCousineau-TRI for platform review, please (on Monday).

N.B. on the low priority tag.

EricCousineau-TRI commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with minor nits.

---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/analysis/test/spring_mass_damper_system.h, line 25 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuKHjSlvjXS38ei3aPR:-KuKHjSmUAppo_Z_jXAz:bjo1whe) ([raw file](https://github.com/robotlocomotion/drake/blob/6f32e66692e9ea75468c5fe68ab3a871010fd573/drake/systems/analysis/test/spring_mass_damper_system.h#L25)):*
> ```objc
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(SpringMassDamperSystem);
> 
>   /// Subclasses must use the protected constructor, not this one.
> ```

BTW This seems like an internal `//` comment - should it be? Or perhaps it should be annotated with `@note`?

---

*[drake/systems/analysis/test/stateless_system.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuKI-TSOaC5SBqU50Fv:-KuKI-TTyVvPmEKz8hPC:b-lobdjd) ([raw file](https://github.com/robotlocomotion/drake/blob/6f32e66692e9ea75468c5fe68ab3a871010fd573/drake/systems/analysis/test/stateless_system.h#L62)):*
> ```objc
> 
>   /// Scalar-converting copy constructor.  See @ref system_scalar_conversion.
>   /// @note this function does not transmogrify the publish callback because
> ```

BTW nit Capitalization on "this"
Also, use of "transmogrify" vs the new term ("scalar-convert"?)?

---

*[drake/systems/analysis/test/stateless_system.h, line 88 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuKKVU_O1hQpFj3_V-M:-KuKKVUa3EunNVP-4ALH:bzdq9a3) ([raw file](https://github.com/robotlocomotion/drake/blob/6f32e66692e9ea75468c5fe68ab3a871010fd573/drake/systems/analysis/test/stateless_system.h#L88)):*
> ```objc
> 
>  private:
>   // Allow different specializations to access each other's private data.
> ```

BTW minor nit Imperative form, should be descriptive?xxxxx

---

*[drake/systems/plants/spring_mass_system/spring_mass_system.h, line 83 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuKJn_R-krtk0gO6gut:-KuKJn_SDKMKCVv_fPip:b-h1gpz0) ([raw file](https://github.com/robotlocomotion/drake/blob/6f32e66692e9ea75468c5fe68ab3a871010fd573/drake/systems/plants/spring_mass_system/spring_mass_system.h#L83)):*
> ```objc
> 
>   /// Construct a spring-mass system with a fixed spring constant and given
>   /// mass.  Subclasses must use the protected constructor, not this one.
> ```

BTW minor xnit Two spaces before sentence begin. Also, doxygen comment is imperative, should be descriptive, "Constructs".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-:-KuKKmirRYkYDY0HFxza:b-8y1vs2)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 8 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/analysis/test/spring_mass_damper_system.h, line 25 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuKHjSlvjXS38ei3aPR:-KuKMbAA4y30_rjY2KSy:b-uoy2u7) ([raw file](https://github.com/robotlocomotion/drake/blob/6f32e66692e9ea75468c5fe68ab3a871010fd573/drake/systems/analysis/test/spring_mass_damper_system.h#L25)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW This seems like an internal `//` comment - should it be? Or perhaps it should be annotated with `@note`?
</blockquote></details>

I think that contracts with subclasses are within the scope of Doxygen comments, for the same reason that protected methods use `///`.

I don't see the reason to use `@note` but will reconsider if you can clarify what you're getting at?

---

*[drake/systems/analysis/test/stateless_system.h, line 62 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuKI-TSOaC5SBqU50Fv:-KuKNZtaVo-lg74HFpnJ:bshjwr9) ([raw file](https://github.com/robotlocomotion/drake/blob/6f32e66692e9ea75468c5fe68ab3a871010fd573/drake/systems/analysis/test/stateless_system.h#L62)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

BTW nit Capitalization on "this"
Also, use of "transmogrify" vs the new term ("scalar-convert"?)?
</blockquote></details>

Done (both).

---

*[drake/systems/plants/spring_mass_system/spring_mass_system.h, line 83 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuKJn_R-krtk0gO6gut:-KuKMzWUMwFG2p_iAbOT:b-z4t674) ([raw file](https://github.com/robotlocomotion/drake/blob/6f32e66692e9ea75468c5fe68ab3a871010fd573/drake/systems/plants/spring_mass_system/spring_mass_system.h#L83)):*
> Two spaces before sentence begin.

Are you saying that a double space after a sentence-ending period is a defect?  I'm not remembering that rule; can you provide a citation?

> Also, doxygen comment is imperative, should be descriptive, "Constructs".

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7031)*
<!-- Sent from Reviewable.io -->

EricCousineau-TRI commented:
Review status: 8 of 10 files reviewed at latest revision, all discussions resolved.

---

*[drake/systems/analysis/test/spring_mass_damper_system.h, line 25 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuKHjSlvjXS38ei3aPR:-KuL5shSKJCNY_s3dmSx:b-wow8tk) ([raw file](https://github.com/robotlocomotion/drake/blob/6f32e66692e9ea75468c5fe68ab3a871010fd573/drake/systems/analysis/test/spring_mass_damper_system.h#L25)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I think that contracts with subclasses are within the scope of Doxygen comments, for the same reason that protected methods use `///`.

I don't see the reason to use `@note` but will reconsider if you can clarify what you're getting at?
</blockquote></details>

For `@note`, I think it'd make a slightly cleaner read in Doxygen to more explicitly delineate the purpose of the statement.

I'm thinking of if I were to open up the Dox, I'd be slightly confused as to why I should even use this constructor at all if this were the only statement.

---

*[drake/systems/plants/spring_mass_system/spring_mass_system.h, line 83 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuKJn_R-krtk0gO6gut:-KuL6pEGbVZjjBBFy2nw:bgcrllc) ([raw file](https://github.com/robotlocomotion/drake/blob/6f32e66692e9ea75468c5fe68ab3a871010fd573/drake/systems/plants/spring_mass_system/spring_mass_system.h#L83)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

> Two spaces before sentence begin.

Are you saying that a double space after a sentence-ending period is a defect?  I'm not remembering that rule; can you provide a citation?

> Also, doxygen comment is imperative, should be descriptive, "Constructs".

Done.
</blockquote></details>

Less a citation, but more so consistency within the file. Most (if not all?) of the other comments below with sentence boundaries on a single line have a single space following the period.

(BTW Typo on my "~x~nit"...)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7031)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: 5 of 10 files reviewed at latest revision, all discussions resolved.

---

*[drake/systems/plants/spring_mass_system/spring_mass_system.h, line 83 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuKJn_R-krtk0gO6gut:-KuLRYAkozkMhGXvIWHl:b2qbwi2) ([raw file](https://github.com/robotlocomotion/drake/blob/6f32e66692e9ea75468c5fe68ab3a871010fd573/drake/systems/plants/spring_mass_system/spring_mass_system.h#L83)):*
<details><summary><i>Previously, EricCousineau-TRI wrote…</i></summary><blockquote>

Less a citation, but more so consistency within the file. Most (if not all?) of the other comments below with sentence boundaries on a single line have a single space following the period.

(BTW Typo on my "~x~nit"...)
</blockquote></details>

Ah, I understand now.  Done throughout.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7031)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
@jwnimmer-tri, you requested that I review this so that I understood and agreed with the changes: yes to the former (although I did indeed need to leverage the associated doc page) and <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> for the latter.

---

Reviewed 5 of 10 files at r1, 4 of 5 files at r2.
Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/systems/analysis/test/stateless_system.h, line 65 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuUf1dZRZzuLQXjIbHy:-KuUf1dZRZzuLQXjIbHz:b4pbdh6) ([raw file](https://github.com/robotlocomotion/drake/blob/f9c540e123bac551572bc8e40c3ec01367b86413/drake/systems/analysis/test/stateless_system.h#L65)):*
> ```objc
>   ///       this is test code for which it is expected that no one will care
>   ///       whether the publish callback survives transmogrification.
>   explicit StatelessSystem(const StatelessSystem<double>& other, int dummy = 0)
> ```

BTW Would you document why the dummy variable is necessary?

---

*[drake/systems/analysis/test/stateless_system.h, line 97 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuUerAyMdmp1ZcfLwK5:-KuUerAyMdmp1ZcfLwK6:b-pqmvht) ([raw file](https://github.com/robotlocomotion/drake/blob/f9c540e123bac551572bc8e40c3ec01367b86413/drake/systems/analysis/test/stateless_system.h#L97)):*
> ```objc
> }  // namespace analysis_test
> 
> namespace scalar_conversion {
> ```

BTW Would you document why this function is necessary?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-:-KuUcpUolla9yWuhWr3t:b-4rwii4)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
With the mention of the weird syntax in StatelessSystem, and more time to think about it, I've found a way to change the StatelessSystem that I think is better.

In general, System classes (and their WitnessFunction helpers) should be wary of using `T`-typed scalars that do not come from a `Context`.  Otherwise, there is no (generic) way for our toolbox to enumerate and configure the partials of the bepoke T -- in this case, the trigger_time offset.  I've reworked the example to use a double for the trigger time, instead of  T.  This simplifies the transmog syntax as a nice bonus.

PTAL.

---

Review status: 9 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/analysis/test/stateless_system.h, line 65 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuUf1dZRZzuLQXjIbHy:-KuUmarPYqma8awt8-yy:blxcwl9) ([raw file](https://github.com/robotlocomotion/drake/blob/f9c540e123bac551572bc8e40c3ec01367b86413/drake/systems/analysis/test/stateless_system.h#L65)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW Would you document why the dummy variable is necessary?
</blockquote></details>

OK The dummy was required because without it, when T = double we'd have the signature of the (standard C++) copy constructor, which conflicts with the DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(StatelessSystem) above.

However, I've removed it here as noted atop the PR.

---

*[drake/systems/analysis/test/stateless_system.h, line 97 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-KuUerAyMdmp1ZcfLwK5:-KuUnr7ZvIK6euruI2Dz:b-ss1bsd) ([raw file](https://github.com/robotlocomotion/drake/blob/f9c540e123bac551572bc8e40c3ec01367b86413/drake/systems/analysis/test/stateless_system.h#L97)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW Would you document why this function is necessary?
</blockquote></details>

OK This was here to say "The scalar conversion can only come-from double" which was required because the trigger_time could only come-from double, because you can't initialize, for example, a double trigger_time from a Symbolic trigger_time.

However, I've removed it here as noted atop the PR.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-:-KuUmtYGpuD5LlzOKwfG:bf5xqti)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
@EricCousineau-TRI or @edrumwri could at least one of you update your review to acknowledge (or discuss) the most recent changes?  Thanks!

EricCousineau-TRI commented:
Done - sorry for the delay!

---

Reviewed 4 of 5 files at r2, 3 of 3 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/7031#-:-Ku_7ERDX2jWtP9togxL:b-exmvl9)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Thanks!  @edrumwri I am happy to adjust for any comments you might still have, and you can still document them here, but I'll do any fixes in a future PR.  Merging now.

edrumwri commented:
Sorry Jeremy- was working on marking the parts in reviewable green and then
got into a conversation with Sean Curtis. I had no further comments.


--------------------------------
Evan Drumwright
Senior Research Scientist
http://positronicslab.github.io
Toyota Research Institute
Palo Alto, CA

On Thu, Sep 21, 2017 at 8:57 AM, Jeremy Nimmer <notifications@github.com>
wrote:

> Merged #7031 <https://github.com/RobotLocomotion/drake/pull/7031>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RobotLocomotion/drake/pull/7031#event-1259690756>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ACHwz5gT02r-hgAi5rXNUMNhl0uRXegBks5skodygaJpZM4PZCGu>
> .
>
